app/lib/github_badges.rb

```diff
@@ -29,40 +68,64 @@ def self.badge_grant!
-
-    if Dir.exists?(path)
-      Rails.logger.info `cd #{path} && git pull`
-    else
-      Rails.logger.info `cd /tmp && git clone #{SiteSetting.github_badges_repo} github_badges`
+    if !Rails.env.test?
+      if Dir.exists?(self.TMP_DIR)
+        Rails.logger.info `cd #{self.TMP_DIR} && git pull`
+      else
+        Rails.logger.info `git clone #{SiteSetting.github_badges_repo} #{self.TMP_DIR}`
+      end
     end
 
-    `cd #{path} && git log --merges --pretty=format:%p --grep='Merge pull request'`.each_line do |m|
-      emails << (`cd #{path} && git log -1 --format=%ce #{m.split(' ')[1].strip}`.strip)
+    Dir.chdir(self.TMP_DIR) do
+      `git log --merges --pretty=format:%p --grep='Merge pull request'`.each_line do |m|
+        emails << (`git log -1 --format=%ce #{m.split(' ')[1].strip}`.strip)
+      end
     end
 
-    email_commits = emails.group_by { |e| e }.map { |k, l| [k, l.count] }.to_h
+    granter = GithubBadges::Granter.new(emails)
+    granter.add_badge(bronze, as_title: false) { |commits| commits >= 1 }
+    granter.add_badge(silver, as_title: true) { |commits| commits >= SiteSetting.github_silver_badge_min_commits }
+    granter.add_badge(gold, as_title: true) { |commits| commits >= SiteSetting.github_gold_badge_min_commits }
+    granter.grant!
 
-    Rails.logger.info "#{email_commits.length} commits found!"
+    committer_badge_grant!
+  end
 
-    User.with_email(email_commits.keys).each do |user|
-      commits = email_commits[user.email]
+  def self.committer_badge_grant!
+    unless bronze = Badge.find_by(name: COMMITER_BADGE_NAME_BRONZE)
+      bronze = Badge.create!(name: COMMITER_BADGE_NAME_BRONZE,
+                             description: 'created a commit',
+                             enabled: false,
+                             badge_type_id: 3)
+    end
 
-      BadgeGranter.grant(bronze, user)
+    unless silver = Badge.find_by(name: COMMITER_BADGE_NAME_SILVER)
+      silver = Badge.create!(name: COMMITER_BADGE_NAME_SILVER,
+                             description: 'created 25 commits',
+                             enabled: false,
+                             badge_type_id: 2)
+    end
 
-      if commits >= SiteSetting.github_silver_badge_min_commits
-        BadgeGranter.grant(silver, user)
-        if user.title.blank?
-          user.update_attributes!(title: silver.name)
-        end
-      end
+    unless gold = Badge.find_by(name: COMMITER_BADGE_NAME_GOLD)
+      gold = Badge.create!(name: COMMITER_BADGE_NAME_GOLD,
+                           description: 'created 1000 commits',
+                           enabled: false,
+                           badge_type_id: 1)
+    end
 
-      if commits >= SiteSetting.github_gold_badge_min_commits
-        BadgeGranter.grant(gold, user)
-        if user.title.blank?
-          user.update_attributes!(title: gold.name)
-        end
+    emails = []
+    Dir.chdir(self.TMP_DIR) do
+      `git log --no-merges --pretty=format:%H`.each_line do |h|
```

Comment:
On very big repositories this could take a while to finish. On disocurse/discourse it took about a minute to finish. Is this OK or should I limit it to the last 10K or 20K commits?

Reply:
I think a minute is ok, no need to put a guard quite yet

Reply:
This takes `345.47 secs` to run on Meta against discourse/discourse. I wonder if we can remove the need for the expensive ruby looping by just running two git commands:

Fetch author email list: 
```
git log --no-merges --format=%ae
```

Fetch co-author email list:
```
git log --no-merges --format=%b | sed -n "s/Co-authored-by: .* \<\(.*\)>/\1/p"
```

These both complete instantaneously for me on discourse/discourse

What do you think @OsamaSayegh?

Reply:
Sounds great to me! Done in https://github.com/discourse/discourse-github/commit/e23b4b892d8e91371664ec4db74aac9196653ce9

Massive improvements:
Before:
```
osama@dev ~/discourse> time rails runner 'GithubBadges.badge_grant!'                                                           master
rails runner 'GithubBadges.badge_grant!'  59.53s user 8.54s system 96% cpu 1:10.57 total
```

After:
```
osama@dev ~/discourse> time rails runner 'GithubBadges.badge_grant!'                                                           master
rails runner 'GithubBadges.badge_grant!'  4.78s user 0.83s system 78% cpu 7.125 total
```

Thanks @davidtaylorhq ❤️